### ISSUE TITLE
Release header mismatch

### ISSUE THREAD

**Issue by yooksi** [joffrey-bion/gradle-github-changelog#30]
Release headers are not properly matched with release text content.

I've provided two changelogs in an attachment below to demonstrate this issue:

- `CHANGELOG1` was generated by `github-changelog-generator`.
- `CHANGELOG2` was generated by `gradle-github-changelog`.

As you can see every header has been moved up for one slot, also notice how `v1.0.0` does not have a link to full changelog.

Attachment: [mismatched-changelogs.zip](https://github.com/joffrey-bion/gradle-github-changelog/files/5935997/mismatched-changelogs.zip)

**Comment by joffrey-bion** [comment]
Thanks for opening this.

Did you use any particular config, or are you using the default?

I always use `futureVersion` myself, so maybe there is indeed a bug when using the "unreleased" default.

**Comment by yooksi** [comment]
This is my plugin configuration:
```
changelog {
    githubUser = 'yooksi'
    githubToken = "${superSecretToken}"
    githubRepository = rootProject.name
    outputFile = file("${projectDir}/CHANGELOG2.md")
}
```
I will try running with `futureVersionTag` and report back.

**Comment by joffrey-bion** [comment]
I tried on a couple of my repos without `futureVersionTag`, and I can't reproduce this issue

**Comment by yooksi** [comment]
I've tried running with `futureVersionTag` alone as well as in combination with `showUnreleased` and the issue persists.

**Comment by joffrey-bion** [comment]
@yooksi I took a look at your repo. The issues and tags dates are correct: the issues were closed after the tag, that's why they appear in the previous section.

**Comment by yooksi** [comment]
> The issues and tags dates are correct: the issues were closed after the tag, that's why they appear in the next section.

I am not sure if that is correct, take a look at commit https://github.com/yooksi/pz-zdoc/commit/3bed3679efa9f6b6155f857ec7322d4c478a6b3e, it was commited on `2021-01-31 17:47:20 +0100` while tag `v2.0.0` was created on `2021-01-31 17:53:11 +0100`. It was closed before `v2.0.0` but is listed under `v2.1.0`.

**Comment by joffrey-bion** [comment]
Mmh maybe I just got unlucky then, lemme take another look.

**Comment by joffrey-bion** [comment]
It's based on issue closing time, not on the commit time. I guess the commit was merged later in this case?

**Comment by yooksi** [comment]
That particular issue that I linked was (like most others) closed with a release merge commit that was tagged soon after.

> Maybe the bug lies here. If the commits are made locally as well as the tag, and everything is pushed later, then the issues will be closed after the tag date.

Isn't that part of normal workflow though? You prepare everything locally and then push it to origin, which will always result in issues being closed after commit and tag dates. How does your release workflow look like?

**Comment by joffrey-bion** [comment]
Personally, I push or merge commits regularly to master, closing issues with the commits. When the time comes to release, I generate, commit and push the changelog, and tag this commit with the version.

**Comment by yooksi** [comment]
> I wonder if there is a way for me to get the information about the closing commit from github.

It would probably be best to see how `github-changelog-generator` handles this.

**Comment by joffrey-bion** [comment]
See this: https://github.com/github-changelog-generator/github-changelog-generator#features-and-advantages-of-this-project

> Manually specify the version that fixed an issue (for cases when the issue's Closed date doesn't match) by giving the issue's milestone the same name as the tag of version

But you didn't use milestones for 2.0.0, right?

**Comment by joffrey-bion** [comment]
> In the meanwhile I will get to work by editing my past tag dates

I could also implement a feature similar to the one I linked above.
This would allow to override the behaviour in bulk.

**Comment by yooksi** [comment]
> But you didn't use milestones for 2.0.0, right?

I have a [milestone](https://github.com/yooksi/pz-zdoc/milestone/1?closed=1) that contains issues closed by `v2.0.0` but I did not specify it before generating.

> It could also be due to poor granularity (if github-changelog-generator only looks at dates, not date+time).

This is probably the most likely explanation.

> I could also implement a feature similar to the one I linked above.
> This would allow to override the behaviour in bulk.

That would definitively solve this issue.

I would suggest you include a note in the plugin `README` that explains this problem and instructions on how to avoid, namely to make sure they create and push their release tag **after** they pushed the release itself to remote repo.

**Comment by yooksi** [comment]
I have updated the date of all tags in `pz-zdoc` repository so that they are dated **after** the closure date of all commits belonging to their respective releases. However, the plugin is still not generating a correct changelog.

Here are the updated tag dates:
```
v1.0.0 Sat Dec 5 23:42:20 2020 +0100
v1.0.1 Sun Dec 6 01:24:17 2020 +0100
v1.0.2 Sun Dec 6 11:13:00 2020 +0100
v2.0.0 Sun Jan 31 18:12:00 2021 +0100
v2.1.0 Fri Feb 5 20:01:00 2021 +0100
```

Examples of misplaced commits:

- This commit should be included in `v1.0.2`
`27292af` - Compiled lua without syntactic sugar
`Sun Dec 6 11:12:00 2020 +0100`

- This commit should be included in `v2.1.0`
`65f1486` - Global functions are not global
`Fri Feb 5 20:00:00 2021 +0100`

**Comment by joffrey-bion** [comment]
Thanks a lot for reporting back, I will look at it ASAP:

Links:
* tag 1.0.2: https://github.com/yooksi/pz-zdoc/releases/tag/v1.0.2
* commit 27292af : https://github.com/yooksi/pz-zdoc/commit/27292af020a794cbe8a44ae04ae8a366dcd31225
* related issue: https://github.com/yooksi/pz-zdoc/issues/1

**Comment by yooksi** [comment]
Just stumbled upon this while reading the `github-changelog-generator` help:

```
--[no-]filter-by-milestone   
Use milestone to detect when issue was resolved. Default is true.
```

Could be an alternative way to determine when issues were resolved.

**Comment by joffrey-bion** [comment]
Yes, this is what I was talking about in [my comment](https://github.com/joffrey-bion/gradle-github-changelog/issues/30#issuecomment-774375718) above.

I think I'm going to add this feature.

**Comment by joffrey-bion** [comment]
Milestone override support was done in https://github.com/joffrey-bion/gradle-github-changelog/issues/33.
